overlord/ifacestate/handlers.go

```diff
@@ -1259,6 +1259,61 @@ func (m *InterfaceManager) doGadgetConnect(task *state.Task, _ *tomb.Tomb) error
 	return nil
 }
 
+// doHotplugUpdateSlot updates static attributes of a hotplug slot for given device.
+func (m *InterfaceManager) doHotplugUpdateSlot(task *state.Task, _ *tomb.Tomb) error {
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	ifaceName, hotplugKey, err := getHotplugAttrs(task)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot get hotplug task attributes: %s", err)
+	}
+
+	var attrs map[string]interface{}
+	if err := task.Get("slot-attrs", &attrs); err != nil {
+		return fmt.Errorf("internal error: cannot get slot-attrs attribute for device %s, interface %s: %s", hotplugKey, ifaceName, err)
+	}
+
+	slot, err := m.repo.SlotForHotplugKey(ifaceName, hotplugKey)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot determine slot for device %s, interface %s: %s", hotplugKey, ifaceName, err)
+	}
+	if slot == nil {
+		return nil
+	}
+
+	stateSlots, err := getHotplugSlots(st)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot obtain hotplug slots: %v", err)
+	}
+
+	conns, err := m.repo.ConnectionsForHotplugKey(ifaceName, hotplugKey)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot determine connections for device %s, interface %s: %s", hotplugKey, ifaceName, err)
+	}
+
+	// hotplug-update-slot is meant to be run as part of a change that first disconnects all slots, then updates the slot and finally
+	// reconnects all connections, so that disconnect- and connect- hooks are run with old and new attributes, respectively. In theory
+	// we should never hit this condition as it should be prevented by conflict logic.
+	if len(conns) > 0 {
+		return fmt.Errorf("internal error: cannot update slot %s while connected", slot.Name)
+	}
+
+	if slotSpec, ok := stateSlots[slot.Name]; ok {
+		slotSpec.StaticAttrs = attrs
+		stateSlots[slot.Name] = slotSpec
+		setHotplugSlots(st, stateSlots)
+
+		// XXX: this is ugly and relies on the slot infos being kept as pointers in the repository
+		slot.Attrs = attrs
```

Comment:
yes, this looks like it should be a method on repo somehow

Reply:
Moved this logic to new `repo.UpdateStaticSlotAttrs(..)` method and simplified `hotplug-update-slot` handler by avoiding extra calls to repo (getting the slot instance itself, checking its connections); all that is now done by `UpdateStaticSlotAttrs`, so it's all happening while the repository is locked.